File: backend/src/main/java/lan/dk/podcastserver/manager/worker/itunes/ItunesFinder.java
package lan.dk.podcastserver.manager.worker.itunes;

import com.github.davinkevin.podcastserver.utils.MatcherExtractor;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.Finder;
import lan.dk.podcastserver.manager.worker.rss.RSSFinder;
import lan.dk.podcastserver.service.JsonService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import javax.validation.constraints.NotEmpty;
import java.util.regex.Pattern;

import static com.github.davinkevin.podcastserver.utils.MatcherExtractor.from;

/**
 * Created by kevin on 12/05/2018
 */
@Service
@RequiredArgsConstructor
public class ItunesFinder implements Finder {
    private static final String ITUNES_API = "https://itunes.apple.com/lookup?id=";
    private static final MatcherExtractor.PatternExtractor ARTIST_ID = from(Pattern.compile(".*id=?([\\d]+).*"));

    private final RSSFinder rssFinder;
    private final JsonService jsonService;

    @Override
    public Podcast find(String url) {
        return ARTIST_ID.on(url)
                .group(1)
                .map(id -> ITUNES_API + id)
                .flatMap(jsonService::parseUrl)
                .map(JsonService.to("results[0].feedUrl", String.class))
                .map(rssFinder::find)
                .getOrElse(Podcast.DEFAULT_PODCAST);
    }

    @Override
    public Integer compatibility(@NotEmpty String url) {
        return url.contains("itunes.apple.com") ? 1 : Integer.MAX_VALUE;
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/manager/worker/itunes/ItunesFinder.kt
package com.github.davinkevin.podcastserver.manager.worker.itunes

import arrow.core.getOrElse
import com.github.davinkevin.podcastserver.utils.MatcherExtractor.Companion.from
import com.github.davinkevin.podcastserver.utils.k
import lan.dk.podcastserver.entity.Podcast
import lan.dk.podcastserver.manager.worker.Finder
import lan.dk.podcastserver.manager.worker.rss.RSSFinder
import lan.dk.podcastserver.service.JsonService
import org.springframework.stereotype.Service

/**
 * Created by kevin on 12/05/2018
 */
@Service
class ItunesFinder(val rssFinder: RSSFinder, val jsonService: JsonService) : Finder {

    override fun find(url: String) =
            ARTIST_ID.on(url)
                    .group(1).k()
                    .map { "$ITUNES_API$it" }
                    .flatMap { jsonService.parseUrl(it).k() }
                    .map { JsonService.to("results[0].feedUrl", String::class.java).apply(it) }
                    .map { rssFinder.find(it) }
                    .getOrElse { Podcast.DEFAULT_PODCAST }

    override fun compatibility(url: String?) =
            if ((url ?: "").contains("itunes.apple.com")) 1
            else Integer.MAX_VALUE

    companion object {
        private const val ITUNES_API = "https://itunes.apple.com/lookup?id="
        private val ARTIST_ID = from(".*id=?([\\d]+).*")
    }
}
